src/System.Private.CoreLib/shared/System/Security/SecureString.cs

```diff
@@ -139,15 +264,41 @@ public void SetAt(int index, char c)
                 {
                     throw new ArgumentOutOfRangeException(nameof(index), SR.ArgumentOutOfRange_IndexString);
                 }
-                Debug.Assert(index <= int.MaxValue / sizeof(char));
 
                 EnsureNotDisposed();
                 EnsureNotReadOnly();
 
-                SetAtCore(index, c);
+                Debug.Assert(_buffer != null);
+
+                SafeBuffer? bufferToRelease = null;
+
+                try
+                {
+                    UnprotectMemory();
+
+                    Span<char> span = AcquireSpan(ref bufferToRelease);
+                    span[index] = c;
+                }
+                finally
+                {
+                    ProtectMemory();
+                    bufferToRelease?.DangerousRelease();
+                }
             }
         }
 
+        private unsafe Span<char> AcquireSpan(ref SafeBuffer? bufferToRelease)
```

Comment:
Not for this PR obviously, but seems like something we should consider exposing on SafeBuffer.

Reply:
It may be nice, but it would come with its own issues:
- `SafeBuffer` length is `ulong` length, but Span can only hold `int`
- One has to be careful about the lifetimes.

SafeBuffer adds very little value over SafeHandle once you start using Span. I have considered not using SafeBuffer here at all, and just use SafeHandle directly. It would add like two extra lines, but eliminated a few ulong/int casts.